### DISCUSSION TITLE
New Architecture migration reports

### DISCUSSION THREAD

**Discussion by brentvatne** [reactwg/react-native-new-architecture#177] [Libraries]
> [!NOTE] 
> tl;dr: The New Architecture is being rolled out in 2024, and now is a great time to test your apps with it enabled. Use this discussion topic to share reports of your experiences enabling the New Architecture in your apps, so we can learn what we have done well, what we need to improve, and help you to work through any issues. 

With the New Architecture rolling out in 2024, the Interop Layer and Bridgeless enabled by default in React Native 0.74, and the [recent and ongoing push to achieve widespread New Architecture support in third-party libraries](https://github.com/reactwg/react-native-new-architecture/discussions/167), now is the time to try enabling it in your app!

You can try the New Architecture in your app even if some of the libraries you use aren't supported. Create a new branch in your repository and remove any of the libraries that aren't compatible, until your app is running. This will give you a good idea of what libraries still need work before you can fully migrate to the New Architecture. We recommend creating issues or pull requests on those libraries' repositories to help them become compatible with the New Architecture.

When you've tried it out, report your findings here! Share the exact React Native version used (0.74.0 or the most recent available at the time of reading is strongly advised), which libraries you used that did not work, and any other problems you observed.

Resources to get started:
- [Enable New Architecture for Apps](https://github.com/reactwg/react-native-new-architecture/blob/main/docs/enable-apps.md)
- [Expo docs on the New Architecture](https://docs.expo.dev/guides/new-architecture/)

**Comment by brentvatne** [comment]
@kadikraman and I tried it on the upcoming React Conf app and it worked well without any changes on both Android and iOS :) here is our **package.json**: https://gist.github.com/brentvatne/6c966e0c5dc2dc857afc893c92904579

- one quirk we notices was that the [headerBlurEffect](https://reactnavigation.org/docs/native-stack-navigator/#headerblureffect) prop (expo-router → react-navigation → react-native-screens) did not work with the New Architecture enabled, I'll follow up by creating a minimal repro and creating an issue for it.
- we have nested touchables to implement bookmarks, so you have a card where the entire card is pressable and then there is another button inside of it, but that second button is not pressable. using gesture-handler
- event bubbling error when using native search bar in screens `(NOBRIDGE) ERROR  Invariant Violation: Event cannot be both direct and bubbling: topFocus`

**Comment by MariuzM** [comment]
Hello, has there been any changes with `Expo51` + `NewArch` and `expo-font`? 

I currently get error when i use `useFonts` + `@gorhom/bottom-sheet` lib, also it only crashes when i use `BottomSheetScrollView` but if don't it works

<img width="1118" alt="image" src="https://github.com/reactwg/react-native-new-architecture/assets/36937039/98922f1f-b053-4bea-9f43-1f9b2070c159">

Recap: 
- if i stop using `expo-fonts` it works
- if i use `expo-fonts` and `BottomSheetScrollView` it crashes
- if i use `expo-fonts` and dont use BottomSheetScrollView` it works

Update: no maybe font is not the problem, i was deleting a lot of my app to test, but once i reverted everything and removed use-fonts also stopped working 
<img width="750" alt="image" src="https://github.com/reactwg/react-native-new-architecture/assets/36937039/09316e02-1831-4dcc-b19e-40c8b2bf9f27">

**Reply by cortinico** [reply]
cc @brentvatne ☝️

**Comment by MariuzM** [comment]
Right here is video, as soon as i make change to BottomSheet then it crashes. 

https://github.com/reactwg/react-native-new-architecture/assets/36937039/cfafa476-3669-48fc-b802-95a03c90310a

Repo to test: https://github.com/MariuzM/expo-starter/blob/main/src/components/BottomSheet.tsx

But if i don't use `BottomSheetScrollView` then it works

**Reply by brentvatne** [reply]
thanks for the repro! we'll look into it

**Comment by murraygunn** [comment]
I upgraded after my phone told me that I had a mismatch (v50 project and v51 on phone). I immediately got an error `Error: Exception in HostFunction: expected 1 arguments, got 0, js engine: hermes`. Some searching turned up similar problems due to the version of `react-native-reanimated`. I upgraded to v3.11.0 as per instructions, but that gave an error that JS was using a different version to the native part, so I downgraded to 3.10.0, which worked, but all my icons disappeared. That can apparently be fixed by adding some config for babel-plugin-module-resolver to babel.config.js, but I'm struggling to get that right.
      plugins: [
        'react-native-reanimated/plugin',
        'babel-plugin-module-resolver', {
            alias: {
                'react-native-vector-icons': '@expo/vector-icons',
            },
        }
      ],
doesn't work.

**Reply by brentvatne** [reply]
hi there! can you share a minimal reproducible example? it sounds like this isn't new architecture related?

**Comment by ibragim3007** [comment]
Stops work after I updated to 51.0

<img width="658" alt="Screenshot 2024-05-08 at 1 39 54 PM" src="https://github.com/reactwg/react-native-new-architecture/assets/40404277/c9e4ae5c-b9e3-4370-8a53-3eaae428cdfc">

<img width="246" alt="Screenshot 2024-05-08 at 1 40 20 PM" src="https://github.com/reactwg/react-native-new-architecture/assets/40404277/7957868a-15a0-4ef8-b58b-331b520eb117">

**Comment by lukaszduszenko89** [comment]
Hi all. I try to start script: npm run reset-project and its not working. 
![Script is not working](https://github.com/reactwg/react-native-new-architecture/assets/131206984/70f117c7-7631-43fa-9a67-9601ecac672d)

**Reply by brentvatne** [reply]
hi there, this discussion is for new architecture, please keep discussions to that. i've noted this issue, we'll look into it and have a fix soon!

**Comment by atilino** [comment]
After updating SDK 51, it compiles correctly but when uploading to "transporter" for iOS I received an email saying the following:

Please correct the following issues and upload a new binary to App Store Connect.

ITMS-90013: Corrupt Image File - The image file 'AppIcon60x60@2x.png' appears to be corrupt.

**Reply by kadikraman** [reply]
This isn't an SDK 51 bug. See [this reply](https://developer.apple.com/forums/thread/751398) from Apple:
> There was an issue processing icons which resulted in this error. The underlying issue has been fixed. Please upload a new build of your app. Sorry for the inconvenience.

I just submitted the app without changes and it worked fine (note, need to rebuild to increment the build number unfortunately).

**Reply by atilino** [reply]
Thank you, I was able to load it without any issues.

**Reply by brentvatne** [reply]
hi there, this is not related to the new architecture. please take such questions to https://chat.expo.dev/

**Comment by Artyom-228** [comment]
Following the transition from SDK 50 to SDK 51, I've noticed that my Tailwind CSS isn't functioning as expected. Without making any adjustments to the Tailwind configuration or similar settings, the issue arose solely from updating my project from SDK 50 to SDK 51.

**Reply by brentvatne** [reply]
hi there, this is not related to the new architecture. please take such questions to https://chat.expo.dev/

**Comment by murraygunn** [comment]
> hi there! can you share a minimal reproducible example? it sounds like this isn't new architecture related?

I agree that the final problem doesn't look like it was directly caused by the upgrade, but you asked for feedback on the process. My upgrade failed because the default version of `react-native-reanimated` included in expo51 doesn't match. That led to a bunch of other issues I'm still working through.

**Reply by murraygunn** [reply]
I've managed to resolve this, but now my navigation drawer menu icon isn't appearing.  I don't believe I have any control over this, and the creators have probably got the same problem in their component that I had in my project. Again, probably not directly related to the new architecture, but the experience isn't good so far. I'd say this wasn't ready for a forced rollout.

**Reply by brentvatne** [reply]
hi there, this is not related to the new architecture. please take such questions to https://chat.expo.dev/

**Comment by Niletix** [comment]
Just encountered many problems, clerk stopped working and I have no idea why, I get a warning in my navigator. And I cannnot downgrade expo in iOS. Forcing me to update the app and having to deal with maybe unfixable problems

**Comment by Hulker** [comment]
Hello, Today I was forced to upgrade since my project on Expo didn't work "Not compatible with the version of Expo Go" (message)  - so I upgraded the Expo SDK and not my project is crashed and I get the below errors **`ERROR  TypeError: Cannot read property 'NativeModule' of undefined, js engine: hermes
 ERROR  Invariant Violation: "main" has not been registered. This can happen if:
* Metro (the local dev server) is run from the wrong folder. Check if Metro is running, stop it and restart it in the current project.
* A module failed to load due to an error and `AppRegistry.registerComponent` wasn't called., js engine: hermes`**  

My project is based on JS 
and now updated with Expo 51

I tried everything possible and not luck  - any idea?? thanks

**Reply by santoswillames** [reply]
I have the same problem

**Reply by brentvatne** [reply]
hi there! please take such questions to https://chat.expo.dev/

**Comment by shamilovtim** [comment]
Hello,

I've enabled newArch in our incubation project [web5 wallet](https://github.com/TBD54566975/web5-wallet)

iOS newArch: works
Android newArch: does not work

iOS bridgeless: does not work (and therefore disabled it)
Android bridgeless: does not work (and therefore disabled it)

Currently Android newArch is blocked by a run time error:
```js
/root/react-native/packages/react-native/ReactCommon/jsi/jsi/jsi-inl.h:187: std::shared_ptr<T> facebook::jsi::Object::getHostObject(Runtime &) const [T = facebook::react::RuntimeSchedulerBinding]: assertion "isHostObject<T>(runtime)" failed
```

This is difficult to debug because I cannot invoke the debugger since the error is so early in the callstack the debugger doesn't have a chance to run and catch it. I've tried to isolate the error by commenting out all code and imports in `App.tsx` and `index.ts` but neither of these stop the error. It seems to come probably from a require, but the debugging experience is painful.

**Reply by MariuzM** [reply]
Do we also need to disable `bridgeless` when using newArch?

**Reply by cortinico** [reply]
> Do we also need to disable bridgeless when using newArch?

Nope you should **not** disable bridgeless.

Also @shamilovtim I don't understand what you mean with:

> iOS newArch: ✅
> iOS bridgeless: ⛔️ (disabled)

Is the first ✅ referring to you enabling the New Architecture, but disabling bridgeless. If so, you're not supposed to disable bridgeless.

If you could provide me with a bit more detail and what is the error, we could look into it and attempt to fix it. A reproducer would also be gold here.

**Reply by shamilovtim** [reply]
Hey @cortinico I provided a link to the repo: https://github.com/TBD54566975/web5-wallet.

> If so, you're not supposed to disable bridgeless.

Why? I thought bridgeless was the 2nd step and required more work from maintainers than only newArch alone. iOS works with newArch only, but not newArch + bridgeless. Android works neither with newArch nor with bridgeless.

Bridgeless seems pretty recent while newarch has been around for a while. I thought bridgeless was an optional final step.

**Reply by shamilovtim** [reply]
> Is the first ✅ referring to you enabling the New Architecture

I'm saying iOS newArch works, but iOS bridgeless does not. 
On Android, neither newArch nor bridgeless work.

**Reply by cortinico** [reply]
> Bridgeless seems pretty recent while newarch has been around for a while. I thought bridgeless was an optional final step.

That's correct but we consider the New Architecture as Fabric + TurboModules + Bridgeless. Ideally you should be testing your libs against Bridgeless only as this helps to reduce the configuration options (which makes things easier for folks to reason about).

**Reply by shamilovtim** [reply]
@cortinico ok if we're not treating bridgeless as an separate optional config then newArch doesn't work in the repro for neither iOS nor Android

**Comment by aretrace** [comment]
### Error recovery for fast refresh 

After enabling new-arch on a development build in a on device iOS Expo project, it appears that the fast refresh feature stops working after an error (e.g., `<Vieww>` instead of `<View>`) is corrected. Consequently, you must press `r` to fully reload the app in order to continue development. This contrasts with the current architecture, where fixing an error allows the app to continue functioning properly with fast refresh still working. The behavior of new-arch can be very frustrating, especially if you are using an editor with auto-save enabled.

**Comment by MD-Aadil-Shafi** [comment]
Hi,
I am using @react-navigation/native for navigation.
after upgrading to expo 51, On Android it's working fine but on ios (my device is iphone 13) it's not working properly. 
once a route is opened and then we switch to different route and then come back to that route it's closing the app.
same with navigate.goBack(), it's also closing the the app even if I'm on next screen.

I've also updated all the packages still this error is there.

**Reply by brentvatne** [reply]
hi there, this topic is for new architecture discussions. please bring other topics to https://chat.expo.dev/

**Comment by ruppinCgroup69** [comment]
Hello,
I have a problem with my app after the upgrade.
as you can see in the video,

https://github.com/reactwg/react-native-new-architecture/assets/158838842/c1b8bb7c-a2d2-474c-aadf-33679e06b987

 as soon as I scan the QR code on my IOS device, the app crashes and closes immediately.
I ccouldnt find a way to fix it. 
Is it a known bug?

**Comment by aretrace** [comment]
When the keyboard opens to enter text in a text input, the KeyboardAvoidingView animation lags slightly behind the animation of the keyboard appearing.

**Comment by szfdiwang** [comment]
When I use the new architecture, the onchange event of TextInput will trigger twice. Don't you even take care of this basic component? My version "react native": "0.74.3", "expo": "~51.0.21",

**Reply by kadikraman** [reply]
Good find! I can reproduce it on Android and I've created an issue for looking into it.
You can track it here  😊 https://github.com/facebook/react-native/issues/45513
